DBZ-3218 Handle duplicate warnings for deprecated options

ani-sha opened the pull request:
https://issues.redhat.com/browse/DBZ-3218

gunnarmorling commented:
@jpechane, looks like you discussed this already with @ani-sha. So is it good to go from your PoV?

rk3rn3r commented on debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:
```diff
@@ -234,7 +234,7 @@ public static DecimalHandlingMode parse(String value, String defaultValue) {
             .withWidth(Width.LONG)
             .withImportance(Importance.LOW)
-            .withValidation(Field::isListOfRegex, RelationalDatabaseConnectorConfig::validateTableExcludeList)
+            .withValidation(Field::isListOfRegex, RelationalDatabaseConnectorConfig::validateTableBlacklist)
```
Not "Blacklist" it should always be Exclude/Include. Other occurrences as well!

gunnarmorling replied:
A separate validation step is needed for old and new option names; hence the separate method and according name.

rk3rn3r commented on debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:
```diff
@@ -636,9 +636,20 @@ public Boolean isFullColummnScanRequired() {
         return getConfig().getBoolean(SNAPSHOT_FULL_COLUMN_SCAN_FORCE);
     }
 
+    private static int validateColumnBlacklist(Configuration config, Field field, Field.ValidationOutput problems) {
```
Here and on the other occurrences field is never used.

gunnarmorling commented on debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:
```diff
@@ -689,9 +710,20 @@ private static int validateTableExcludeList(Configuration config, Field field, V
     }
 
+    private static int validateSchemaBlacklist(Configuration config, Field field, Field.ValidationOutput problems) {
+        String whitelist = Configuration.getFallbackStringProperty(config, SCHEMA_INCLUDE_LIST, SCHEMA_WHITELIST);
```
Why is it using the fallback approach here? Shouldn't it _only_ validate the legacy one?

ani-sha replied:
@gunnarmorling I didn't quite follow your question, `SCHEMA_WHITELIST` is the deprecated option and hence calls the fallback method to log a warning. That is the main idea. Isn't?

rk3rn3r replied:
We need this at the moment for blacklist properties. But agreed, it's a bit weird!
More specifically we use that to trigger the log message (on a different field) when both options (`*.exclude.list and `*.whitelist`) were set.
Although this works I was thinking about defining another (internal) field that does the validation of all the other fields in one place. The question here is: how to properly handle multi-field validation properly (other options?). This is definitely the correct way with the current implementation.

gunnarmorling reviewed (commented):
@ani-sha, one question inline.

rk3rn3r commented:
@ani-sha there's some duplicate code now.
@gunnarmorling do you think it's worth the effort to create a method that takes the 4 field (legacy-allow, new-allow, legacy-deny, new-deny, error-msg) as parameters and call that one from the different static implementations per field?

ani-sha commented:
@jpechane @gunnarmorling Any suggestions on how to proceed with this? Separating the blacklist properties has duplicate code now and no real impact.

gunnarmorling commented:
So I'm not too concerned about duplication here; keep in mind we're going to phase out the legacy fields eventually anyways, so no need to spend much effort "optimizing" this. So if we all agree this PR is doing what it should be doing from a functional perspective, I'd suggest we merge it. Any objections, @ani-sha, @rk3rn3r?

That said, @ani-sha, I'd still like to see an example of where there actually was a duplication _without_ this PR. Any pointer to a test showing that on the current master branch would be great.

rk3rn3r commented:
@gunnarmorling from functional POV it should do the job... in a weird way though. :laughing: 
But good to go so far!

ani-sha commented:
Hi @gunnarmorling, So I have added a separate method for logging, and you can find the differences in `PostgresConnectorIT.shouldCloseTxAfterTypeQuery` [(look for schema/table.include.list](https://github.com/debezium/debezium/runs/2128382460?check_suite_focus=true )). 

The reason why I added this separate method - because the earlier piece of code didn't actually do anything, my apologies I had been looking for the wrong logs. 

But IMO this is not a great way and the code has been quite repetitive. (either we drop this or think of an alternative.)

gunnarmorling commented on debezium-core/src/main/java/io/debezium/config/Configuration.java:
```diff
@@ -2125,6 +2125,15 @@ default String getFallbackStringProperty(Field newProperty, Field oldProperty) {
      * @return the evaluated value
      */
     static String getFallbackStringProperty(Configuration config, Field newProperty, Field oldProperty) {
+        if (null != config.getString(oldProperty.name()) && null != config.getString(newProperty.name())) { // both are set
```
```suggestion
        if (config.hasKey(oldProperty.name()) && config.hasKey(newProperty.name())) { // both are set
```

This is a bit more descriptive of what we want to do here. Same in the new method below.

rk3rn3r replied:
Is this the same meaning? `getString()` will retutn a non-null default value, if set for properties that have non-null default values. Agree, the question is what do we want to achieve here?

gunnarmorling commented on debezium-core/src/main/java/io/debezium/config/Configuration.java:
```diff
@@ -2125,6 +2125,15 @@ default String getFallbackStringProperty(Field newProperty, Field oldProperty) {
      * @return the evaluated value
      */
     static String getFallbackStringProperty(Configuration config, Field newProperty, Field oldProperty) {
```
This method seems only to be used by the instance method right above. So let's inline this one, leaving only the instance method behind.

rk3rn3r replied:
I think this static method is used on many other places (that's why we also need a logging and non-logging version of the code. For example see Filters, RelationalTableFilters, *ConnectoConfig implementations, etc, isn't it?

gunnarmorling commented on debezium-core/src/main/java/io/debezium/config/Configuration.java:
```diff
@@ -2125,6 +2125,15 @@ default String getFallbackStringProperty(Field newProperty, Field oldProperty) {
+                () -> config.getString(oldProperty));
+    }
+
+    static String logFallbackStringProperty(Configuration config, Field newProperty, Field oldProperty) {
```
Having removed the static method above, I'd suggest to rename this one into `getFallbackStringPropertyWithWarning()`, also adding a JavaDoc comment to the method (no need for param/return value tags, as those are evident -- we overdid this a bit in the past --).

gunnarmorling commented on debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:
```diff
@@ -636,9 +636,20 @@ public Boolean isFullColummnScanRequired() {
+        String whitelist = Configuration.logFallbackStringProperty(config, COLUMN_EXCLUDE_LIST, COLUMN_BLACKLIST);
+
+        if (whitelist != null && blacklist != null) {
+            problems.accept(COLUMN_BLACKLIST, blacklist, COLUMN_INCLUDE_LIST_ALREADY_SPECIFIED_ERROR_MSG);
```
Can we split up `COLUMN_INCLUDE_LIST_ALREADY_SPECIFIED_ERROR_MSG` into two, one for include list and one for the legacy term? The reason being, that the legacy term shouldn't show up for a wrong usage of the new options. Same for the other similar messages.

gunnarmorling commented on debezium-core/src/test/java/io/debezium/config/ConfigurationTest.java:
```diff
@@ -70,7 +70,7 @@ public void shouldThrowValidationOnDuplicateOldColumnFilterConfigurationOld() {
                 .build();
 
-        List<String> errorMessages = config.validate(Field.setOf(COLUMN_EXCLUDE_LIST)).get(COLUMN_EXCLUDE_LIST.name()).errorMessages();
+        List<String> errorMessages = config.validate(Field.setOf(COLUMN_BLACKLIST)).get(COLUMN_BLACKLIST.name()).errorMessages();
```
Is this change actually needed?

ani-sha replied:
yes, otherwise the test would fail!

gunnarmorling reviewed (commented):
@ani-sha, so overall I think this change is good. A few comments inline about some details. Thanks a lot for fighting through this really complex logic here!

ani-sha commented:
Rework and clean-up done. Added the same set of changes for `MongoDbConnectorConfig` as well.

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Filters.java:
```diff
@@ -32,8 +32,8 @@
     public Filters(Configuration config) {
-        String dbIncludeList = config.getFallbackStringProperty(MongoDbConnectorConfig.DATABASE_INCLUDE_LIST, MongoDbConnectorConfig.DATABASE_WHITELIST);
-        String dbExcludeList = config.getFallbackStringProperty(MongoDbConnectorConfig.DATABASE_EXCLUDE_LIST, MongoDbConnectorConfig.DATABASE_BLACKLIST);
+        String dbIncludeList = config.getFallbackStringProperty(config, MongoDbConnectorConfig.DATABASE_INCLUDE_LIST, MongoDbConnectorConfig.DATABASE_WHITELIST);
```
@gunnarmorling Should we really call an instance method and pass the instance as parameter to that method? 
That's why I added the instance method with a cleaner parameter list and the static one which actually handled the code.

gunnarmorling replied:
Hum, no. I was expecting `config.getFallbackStringProperty( MongoDbConnectorConfig.DATABASE_INCLUDE_LIST, MongoDbConnectorConfig.DATABASE_WHITELIST);`. Passing the config instance again to an instance method doesn't make sense.

rk3rn3r commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java:
```diff
@@ -34,8 +34,10 @@
 
-    protected static final String COLLECTION_INCLUDE_LIST_ALREADY_SPECIFIED_ERROR_MSG = "\"collection.include.list\" or \"collection.whitelist\" is already specified";
-    protected static final String DATABASE_INCLUDE_LIST_ALREADY_SPECIFIED_ERROR_MSG = "\"database.include.list\" or \"database.whitelist\" is already specified";
+    protected static final String COLLECTION_EXCLUDE_LIST_ALREADY_SPECIFIED_ERROR_MSG = "\"collection.exclude.list\" is already specified";
```
When we have both `*.blacklist` and `*.exclude.list` are set we use the `*.exclude.list` value.
I think there shouldn't be an error on the `*.exclude.list` field, but and only on the `*.blacklist` field, as this value is discarded. WDYT?

gunnarmorling commented:
Closing in favour of #2233. Thanks, @ani-sha!
